controllers/oscmachine_keypair_controller.go

```diff
@@ -0,0 +1,119 @@
+	nodeSpec := machineScope.GetNode()
+	if nodeSpec.KeyPair.Name == "" {
+		nodeSpec.SetKeyPairDefaultValue()
+		keypairSpec = &nodeSpec.KeyPair
+	} else {
+		keypairSpec = machineScope.GetKeypair()
+	}
+
+	keypairName := keypairSpec.Name
+	keypairTagName, err := compute.ValidateKeypairName(keypairName)
+	if err != nil {
+		return keypairTagName, err
+	}
+
+	publicKey := keypairSpec.PublicKey
+	_, err = security.ValidateKeypairPublic(publicKey)
+	if err != nil {
+		return keypairTagName, err
+	}
+	return "", nil
+}
+
+// getKeyPairResourceId return the keypairName from the resourceMap base on resourceName (tag name + cluster uid)
+func getKeyPairResourceId(resourceName string, machineScope *scope.MachineScope) (string, error) {
+	keypairRef := machineScope.GetKeypairRef()
+	if keypairName, ok := keypairRef.ResourceMap[resourceName]; ok {
+		return keypairName, nil
+	} else {
+		return "", fmt.Errorf("%s does not exist", resourceName)
+	}
+}
+
+// reconcileKeypair reconcile the keypair of the machine
+func reconcileKeypair(ctx context.Context, machineScope *scope.MachineScope, keypairSvc security.OscKeyPairInterface) (reconcile.Result, error) {
+	machineScope.Info("Create Keypair or add existing one")
+	var keypairSpec *infrastructurev1beta1.OscKeypair
+	keypairSpec = machineScope.GetKeypair()
+	keypairRef := machineScope.GetKeypairRef()
+	keypairName := keypairSpec.Name + "-" + machineScope.GetUID()
+
+	if len(keypairRef.ResourceMap) == 0 {
+		keypairRef.ResourceMap = make(map[string]string)
+	}
+	if keypairSpec.ResourceId != "" {
+		keypairRef.ResourceMap[keypairName] = keypairSpec.ResourceId
+	}
+
+	if keypair, err := keypairSvc.GetKeyPair(keypairName); err != nil {
+		return reconcile.Result{}, err
+	} else if keypair == nil || keypairSpec.ResourceId == "" {
+		goto CREATE
+	}
+
+	if keypairSpec.PublicKey != "" {
+		security.ValidateKeypairPublic(keypairSpec.PublicKey)
+		_, err := security.ValidateKeypairPublic(keypairSpec.PublicKey)
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+	}
+
+CREATE:
+	keypair, err := keypairSvc.CreateKeyPair(keypairSpec, keypairName)
+	if err != nil {
+		return reconcile.Result{}, fmt.Errorf("%w Can not create keypair for OscCluster %s/%s", err, machineScope.GetNamespace(), machineScope.GetName())
+	}
+
+	machineScope.Info("### Get keypair ###", "keypair", keypair)
+	keypairSpec.ResourceId = keypair.GetKeypairName()
+	keypairRef.ResourceMap[keypairName] = keypair.GetKeypairName()
+	return reconcile.Result{}, nil
+}
+
+// reconcileDeleteKeypair reconcile the destruction of the keypair of the machine
+func reconcileDeleteKeypair(ctx context.Context, machineScope *scope.MachineScope, keypairSvc security.OscKeyPairInterface) (reconcile.Result, error) {
+	machineScope.Info("Delete Key pair")
+	oscmachine := machineScope.OscMachine
+	var keypairSpec *infrastructurev1beta1.OscKeypair
+	nodeSpec := machineScope.GetNode()
+	if nodeSpec.KeyPair.Name == "" {
+		nodeSpec.SetKeyPairDefaultValue()
```

Comment:
In this condition, you would regenerate a publicKey, which is quite strange.

Let's try not to generate the public key here